docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,805 @@
+
+The {plugins}/discovery-ec2.html[EC2 discovery plugin] adds a hosts provider
+that uses the https://github.com/aws/aws-sdk-java[AWS API] to find a list of
+seed nodes.
+
+[float]
+[[azure-classic-hosts-provider]]
+===== Azure Classic hosts provider
+
+The {plugins}/discovery-azure-classic.html[Azure Classic discovery plugin] adds
+a hosts provider that uses the Azure Classic API find a list of seed nodes.
+
+[float]
+[[gce-hosts-provider]]
+===== Google Compute Engine hosts provider
+
+The {plugins}/discovery-gce.html[GCE discovery plugin] adds a hosts provider
+that uses the GCE API find a list of seed nodes.
+
+[float]
+==== Discovery settings
+
+Discovery operates in two phases: First, each node probes the addresses of all
+known master-eligible nodes by connecting to each address and attempting to
+identify the node to which it is connected. Secondly it shares with the remote
+node a list of all of its known master-eligible peers and the remote node
+responds with _its_ peers in turn. The node then probes all the new nodes that
+it just discovered, requests their peers, and so on.
+
+If the node is not master-eligible then it continues this discovery process
+until it has discovered an elected master node. If no elected master is
+discovered then the node will retry after `discovery.find_peers_interval` which
+defaults to `1s`.
+
+If the node is master-eligible then it continues this discovery process until it
+has either discovered an elected master node or else it has discovered enough
+masterless master-eligible nodes to complete an election. If neither of these
+occur quickly enough then the node will retry after
+`discovery.find_peers_interval` which defaults to `1s`.
+
+The discovery process is controlled by the following settings.
+
+`discovery.find_peers_interval`::
+
+    Sets how long a node will wait before attempting another discovery round.
+    Defaults to `1s`.
+
+`discovery.request_peers_timeout`::
+
+    Sets how long a node will wait after asking its peers again before
+    considering the request to have failed. Defaults to `3s`.
+
+`discovery.probe.connect_timeout`::
+
+    Sets how long to wait when attempting to connect to each address. Defaults
+    to `3s`.
+
+`discovery.probe.handshake_timeout`::
+
+    Sets how long to wait when attempting to identify the remote node via a
+    handshake. Defaults to `1s`.
+
+`discovery.cluster_formation_warning_timeout`::
+
+    Sets how long a node will try to form a cluster before logging a warning
+    that the cluster did not form. Defaults to `10s`.
+
+If a cluster has not formed after `discovery.cluster_formation_warning_timeout`
+has elapsed then the node will log a warning message that starts with the phrase
+`master not discovered` which describes the current state of the discovery
+process.
+
+[float]
+[[modules-discovery-bootstrap-cluster]]
+=== Bootstrapping a cluster
+
+Starting an Elasticsearch cluster for the very first time requires the initial
```

Comment:
I think it will be good to work something into this that tells people that this is not needed for an existing cluster as we read this information from the data folder.

Reply:
Agreed, see 4180e0093a12f4900cf55a04e50aa5f6f58f2d4d.